instrumentation/opentelemetry-instrumentation-sklearn/src/opentelemetry/instrumentation/sklearn/__init__.py

```diff
@@ -0,0 +1,738 @@
+
+.. code-block:: python
+
+    from opentelemetry.instrumentation.sklearn import SklearnInstrumentor
+    from sklearn.datasets import load_iris
+    from sklearn.ensemble import RandomForestClassifier
+    from sklearn.model_selection import train_test_split
+    from sklearn.pipeline import Pipeline
+
+    X, y = load_iris(return_X_y=True)
+    X_train, X_test, y_train, y_test = train_test_split(X, y)
+
+    model = Pipeline(
+        [
+            ("class", RandomForestClassifier(n_estimators=10)),
+        ]
+    )
+
+    model.fit(X_train, y_train)
+
+    SklearnInstrumentor().instrument_estimator(model)
+
+"""
+import logging
+import os
+from functools import wraps
+from importlib import import_module
+from inspect import isclass
+from pkgutil import iter_modules
+from typing import Callable, Dict, List, MutableMapping, Sequence, Type, Union
+
+from sklearn.base import BaseEstimator
+from sklearn.pipeline import FeatureUnion, Pipeline
+from sklearn.tree import BaseDecisionTree
+from sklearn.utils.metaestimators import _IffHasAttrDescriptor
+
+from opentelemetry.instrumentation.instrumentor import BaseInstrumentor
+from opentelemetry.instrumentation.sklearn.version import __version__
+from opentelemetry.trace import get_tracer
+from opentelemetry.util.types import Attributes
+
+logger = logging.getLogger(__name__)
+
+
+def implement_spans(
+    func: Callable,
+    estimator: Union[BaseEstimator, Type[BaseEstimator]],
+    attributes: Attributes = None,
+):
+    """Wrap the method call with a span.
+
+    Args:
+        func: A callable to be wrapped in a span
+        estimator: An instance or class of an estimator
+        attributes: Attributes to apply to the span
+
+    Returns:
+        The passed function wrapped in a span.
+    """
+    if isclass(estimator):
+        name = estimator.__name__
+    else:
+        name = estimator.__class__.__name__
+    logger.debug("Instrumenting: %s.%s", name, func.__name__)
+
+    @wraps(func)
+    def wrapper(*args, **kwargs):
+        with get_tracer(__name__, __version__).start_as_current_span(
+            name="{cls}.{func}".format(cls=name, func=func.__name__),
+            attributes=attributes,
+        ):
+            return func(*args, **kwargs)
+
+    return wrapper
+
+
+def implement_spans_delegator(obj: _IffHasAttrDescriptor):
+    """Wrap the descriptor's fn with a span.
+
+    Args:
+        obj: An instance of _IffHasAttrDescriptor
+    """
+    # Don't instrument inherited delegators
+    if hasattr(obj, "_otel_original_fn"):
+        logger.debug("Already instrumented: %s", obj.fn.__qualname__)
+        return
+
+    def implement_spans_get(func: Callable):
+        @wraps(func)
+        def wrapper(*args, **kwargs):
+            with get_tracer(__name__, __version__).start_as_current_span(
+                name=func.__qualname__
+            ):
+                return func(*args, **kwargs)
+
+        return wrapper
+
+    logger.debug("Instrumenting: %s", obj.fn.__qualname__)
```

Comment:
Is `fn` an attribute of all `_IffHasAttrDescriptor`?

Reply:
Yes, it is a required init arg and is the method which is being delegated. https://github.com/scikit-learn/scikit-learn/blob/0fb307bf39bbdacd6ed713c00724f8f871d60370/sklearn/utils/metaestimators.py#L79

Here is the decorator: https://github.com/scikit-learn/scikit-learn/blob/0fb307bf39bbdacd6ed713c00724f8f871d60370/sklearn/utils/metaestimators.py#L144